Simplified/AxisService/NYPLAxisService.swift

```diff
@@ -72,253 +56,70 @@ import Foundation
-  static private let writeFailureSummary = "AXIS: Failed writing item to specified write URL"
-  static private let failedDownloadingItemSummary = "AXIS: Failed downloading item"
-  static private let failedDownloadingAssetsFromPackageSummary = "AXIS: Failed downloading content from pacakge.opf"
-  
   // MARK: - Initialization
-  init(delegate: NYPLBookDownloadBroadcasting?,
-       isbn: String,
-       bookVaultId: String,
-       fileURL: URL,
+  init(axisItemDownloader: NYPLAxisItemDownloading,
        book: NYPLBook,
-       deviceInfoProviding: NYPLDeviceInfoProviding,
-       packageEndpointproviding: NYPLAxisPackageEndpointProviding,
-       packagePathProviding: NYPLAxisPackagePathPrefixProviding,
-       axisKeysProviding: NYPLAxisKeysProviding,
-       assetWriter: NYPLAssetWriting,
-       cypher: NYPLRSACryptographing?,
        dedicatedWriteURL: URL,
-       downloader: NYPLAxisContentDownloading?) {
+       delegate: NYPLBookDownloadBroadcasting?,
+       licenseService: NYPLAxisLicenseHandling,
+       metadataDownloader: NYPLAxisMetadataContentHandling,
+       packageDownloader: NYPLAxisPackageHandling) {
     
-    self.axisKeysProviding = axisKeysProviding
-    self.isbn = isbn
-    self.fileURL = fileURL
-    self.delegate = delegate
-    self.bookVaultId = bookVaultId
-    self.baseURL = axisKeysProviding.baseURL.appendingPathComponent(isbn)
+    self.axisItemDownloader = axisItemDownloader
     self.book = book
-    self.deviceInfoProviding = deviceInfoProviding
-    self.packagePathProviding = NYPLAxisPackagePathPrefixProvider()
-    self.dispatchGroup = DispatchGroup()
-    self.assetWriter = assetWriter
-    self.packageEndpointproviding = packageEndpointproviding
-    self.cypher = cypher
     self.dedicatedWriteURL = dedicatedWriteURL
-    self.downloader = downloader
+    self.delegate = delegate
+    self.licenseService = licenseService
+    self.metadataDownloader = metadataDownloader
+    self.packageDownloader = packageDownloader
+    super.init()
+    weak var weakSelf = self
+    self.axisItemDownloader.delegate = weakSelf
   }
   
   /// Fulfill AxisNow license. Notifies NYPLBookDownloadBroadcasting upon completion or failure.
-  
   /// - Parameters:
   ///   - downloadTask: downloadTask download task
   @objc func fulfillAxisLicense(downloadTask: URLSessionDownloadTask) {
     DispatchQueue.global(qos: .utility).async {
-      self.downloadLicense(forBook: self.book)
-      self.downloadEncryption(forBook: self.book)
-      self.downloadContainer(forBook: self.book)
-      self.downloadPackage(forBook: self.book)
-      self.downloadAssetsFromPackage(forBook: self.book)
-      
-      self.dispatchGroup.notify(queue: .global(qos: .utility)) {
-        _ = self.delegate?.replaceBook(self.book,
-                                       withFileAtURL: self.dedicatedWriteURL,
-                                       forDownloadTask: downloadTask)
+      self.downloadAndValidateLicense()
```

Comment:
these tasks will be running concurrently, are there concurrency issues in doing that? I suppose as long as `downloadMetadataContent` and `downloadPackage` are thread safe we are good?

Reply:
Each method in license service depends on dispatch group. Before starting, it waits for the dispatch group to finish doing whatever it was doing, enters the dispatch group, and doesn't leave it until it's done or encounters an error. This ensures that no two methods run at the same time.

Reply:
All the methods in packageService and metadataService are controlled by dispatchGroup. Before executing, they wait for the dispatchGroup to finish whatever it was doing, then enter the dispatchGroup, and only leave it upon success or failure. This ensures that package service doesn't actually start until metadataService is finished.